presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java

```diff
@@ -0,0 +1,169 @@
+import com.facebook.presto.operator.SpillContext;
+import com.facebook.presto.operator.exchange.LocalPartitionGenerator;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableList;
+import com.google.common.io.Closer;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.function.IntPredicate;
+import java.util.function.Supplier;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Verify.verify;
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class GenericPartitioningSpiller
+        implements PartitioningSpiller
+{
+    private final List<Type> types;
+    private final int partitionsCount;
+    private final PageBuilder[] pageBuilders;
+    private final LocalPartitionGenerator partitionGenerator;
+    private final Supplier<SpillContext> spillContextSupplier; // TODO use. Why Supplier ?
+    private final SingleStreamSpiller[] spillers;
+    private final Closer closer = Closer.create();
+    private boolean readingStarted;
+    private Set<Integer> spilledPartitions = new HashSet<>();
+
+    public GenericPartitioningSpiller(
+            List<Type> types,
+            LocalPartitionGenerator partitionGenerator,
+            int partitionsCount,
+            Supplier<SpillContext> spillContextSupplier,
+            AggregatedMemoryContext memoryContext,
+            SingleStreamSpillerFactory spillerFactory)
+    {
+        this.partitionsCount = partitionsCount;
+        this.types = requireNonNull(types, "types is null");
+        this.spillers = new SingleStreamSpiller[partitionsCount];
+        this.pageBuilders = new PageBuilder[partitionsCount];
+        this.partitionGenerator = requireNonNull(partitionGenerator, "partitionGenerator is null");
+        this.spillContextSupplier = requireNonNull(spillContextSupplier, "spillContextSupplier is null");
+
+        for (int partition = 0; partition < partitionsCount; partition++) {
+            pageBuilders[partition] = new PageBuilder(types);
+            SingleStreamSpiller spiller = spillerFactory.create(types, spillContextSupplier.get(), memoryContext.newLocalMemoryContext());
+            spillers[partition] = spiller;
+            closer.register(spiller);
+        }
+    }
+
+    @Override
+    public synchronized Iterator<Page> getSpilledPages(int partition)
+    {
+        readingStarted = true;
+        getFutureValue(flush(partition));
```

Comment:
It would be better to just check in this method that spill is finished (`flush(partition).isDone() == true`). And throw exception otherwise.
Waiting for spill to complete may be long and contract I propose is in line with one from `SingleStreamSpiller`.